### PR TITLE
UPSTREAM: <carry>: garbage collector monitors syncing

### PR DESCRIPTION
**Pull request by astefanutti**
Adds the ability to update garbage collector monitors for integrators.

Required for kcp-dev/kcp#2112.

### CODE REVIEW

**Comment by ncdc** [comment]
Are you going to do an AST generator for this?

**Comment by astefanutti** [comment]
> Are you going to do an AST generator for this?

Yes, just preferred that you have a look at https://github.com/kcp-dev/kcp/pull/2112/commits/2e900e70c34b0abdc1d41f50231ff85dc15f9d60 before going forward.

**Comment by astefanutti** [comment]
@ncdc I've just added the AST-based generator. Let me know if there is more to do.

**Comment by ncdc** [comment]
@astefanutti thanks! Could you please create commits in this order?

1. The AST generator by itself
2. The unmodified output of running the generator (compare to commit 94dc4f4f0f0)
3. The modifications to the patch file (compare to commit bcfaeb030a9)

**Comment by astefanutti** [comment]
@ncdc thanks a lot. I've just re-pushed with commits articulated according to your instructions. Let me know if that doesn't meet your expectations.

**Comment by ncdc** [comment]
This is great! 1 last request: can you include the clusterName in any/all log messages & errors? See e.g. https://github.com/kcp-dev/kubernetes/blob/6cfb7d485cbfea2b7eb6ff541af8057d3f3e7204/pkg/controller/resourcequota/resource_quota_controller_patch.go#L99 where `rq.clusterName` comes from https://github.com/kcp-dev/kubernetes/blob/6cfb7d485cbfea2b7eb6ff541af8057d3f3e7204/pkg/controller/resourcequota/resource_quota_controller.go#L126

**Comment by astefanutti** [comment]
@ncdc I've just re-pushed with an extra commit that adds the cluster name to all log statements and errors.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 1]
```diff
@@ -78,11 +80,33 @@ type GarbageCollector struct {
...
 	workerLock sync.RWMutex
+
+	// kcp
+	clusterName logicalcluster.Name
 }
 
 var _ controller.Interface = (*GarbageCollector)(nil)
 var _ controller.Debuggable = (*GarbageCollector)(nil)
 
+// NewClusterAwareGarbageCollector creates a new GarbageCollector for the provided logical cluster.
```
Comment (ncdc) [inline pkg/controller/garbagecollector/garbagecollector.go]: This could go in a separate _patch.go file to minimize rebase conflicts - wdyt?

Comment (astefanutti) [reply pkg/controller/garbagecollector/garbagecollector.go]: Absolutely good idea! Re-pushed with 52fa5f286ecb1c1f273ef4ecd06043d5531d4d02 that move the factory method into the patch file.

File: pkg/controller/garbagecollector/garbagecollector.go [hunk 2]
```diff
@@ -146,16 +151,16 @@ func (gc *GarbageCollector) Run(ctx context.Context, workers int) {
 	defer gc.attemptToOrphan.ShutDown()
 	defer gc.dependencyGraphBuilder.graphChanges.ShutDown()
 
-	klog.Infof("Starting garbage collector controller")
-	defer klog.Infof("Shutting down garbage collector controller")
+	klog.Infof("%s: Starting garbage collector controller", gc.clusterName)
```
Comment (stevekuznetsov) [inline pkg/controller/garbagecollector/garbagecollector.go:154]: Not critical, but a slightly easier approach would be to make a commit that moves this controller to use contextual logging, which helps the upstream efforts and can land in k8s quickly, then simply add the cluster name to the logger we pass to the context that we use to `Run()` the controller

Comment (ncdc) [reply pkg/controller/garbagecollector/garbagecollector.go:154]: Would like to proceed now. We can revert this later and adapt on top of contextual logging when my PR lands upstream.

**Review by ncdc — APPROVED** [review]

**Pull request merged**